### PR TITLE
stm32 unify DMA WIP RFC

### PR DESCRIPTION
**Pull request by lebrush**
This tries to unify the DMA controllers for stm32 platforms. There are two identified flavours of DMA, the one which uses channels and the one which uses streams. Also, the bus is usually the same for all using streams and for all using channels (except for the L4...). The biggest difference part is the interrupt lines for the channel based ones.

This is a WIP. I tried to implement them as a single dma.c but also as different files `dma_channel.c` and `dma_stream.c` however I think it's cleaner to have them in a single file so we are sure it's always compiled in.

The main goal is to provide a platform agnostic transfer function `dma_transfer` which could be used by any peripheral or application. See the test application (currently `hello-world/main.c`) to have a better idea of how to use this.

It has been tested with an F0, F1, L0 and L1 (all channel based). F4 (stream based) will be tested soon.

Note that this PR is quite dirty and needs cleanup.

### CODE REVIEW

File: cpu/stm32_common/include/dma_stream.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+{
+    return (dma & 0xff);
+}
+
+/**
+ * @brief Get the channel number of a DMA unit
+ */
+static inline int dma_chnum(dma_t dma)
+{
+    return (dma >> 16);
```
Comment (lebrush) [inline cpu/stm32_common/include/dma_stream.h]: This should be a `>> 8`.

**Comment by vincent-d** [comment]
A first remark after a really quick review. We already tried to implement a dma driver, for stm32f2 (#5584) and one of the interest of a shared driver instead of dma-related code in each periph driver is to manage the shared streams. For instance on stm32f2, the stream 3 is shared by USART3_TX ans SPI2_RX. The channel must then be locked during a transfer and the isr cannot be implemented by one of the drivers because, obviously, it's shared.

But, if I understood it correctly, the proposal in this PR, is that the user code implements the isr. Which won't work for shared stream. Did I misunderstood here?

The main drawback of the proposal we made was it allocated two mutexes by stream, even if they were not used.

Thanks for working on that, anyway :)

**Comment by lebrush** [comment]
Thanks for the feedback. You are absolutely right, this would not work for shared streams, except if the user implements it. At the moment it offers only a very basic usage and the user has to take care of the isrs. Maybe it could be extended to handle shared streams in a later iteration. Good point!

**Comment by lebrush** [comment]
@vincent-d I will take a different approach and try to accommodate what you suggested

**Pull request closed without merging**